worker/uniter/runner/context/context.go

```diff
@@ -483,6 +486,17 @@ func (ctx *HookContext) ConfigSettings() (charm.Settings, error) {
 	return result, nil
 }
 
+func (ctx *HookContext) GoalState() (string, error) {
+	if ctx.goalState == "" {
```

Comment:
I am not sure that caching here is helpful. We cache things like private address because they are not expected to change. Goal state is expected to potentially change with each call.